### PR TITLE
Perform structural Validations

### LINKED ISSUE #298: Validator Package handling Structural Checks
_From @cmh2166 on March 13, 2018 18:23_

- [x] if structural metadata exists, check that it is valid, i.e.: 
    - [x] `File` is contained only by a `Fileset` resource & only contained by 1 `Fileset`
    - [x] `Fileset` is contained only by a `Object` resource
    - [x]  `Fileset` is only contained by 1 `Object` (Unnecessary to validate, because it's a string in the MAP.)
    - [ ] ~~`Object` can only contain `Filesets`~~ (Deprecated)
    - [x] `Object` can only have `Object` as members
    - [x] `Collection` can only have members that are `Collection` or `Object`
    - [x] `structural.hasMemberOrders.members` must be in `structural.hasMembers`

Defer to structural checks documented here: https://github.com/sul-dlss-labs/taco/wiki/TACO-Internal-Processing-Steps#deposit-resource

_Copied from original issue: sul-dlss/sdr3-planning#29_

### PR DESCRIPTION
**Pull request by jcoyne**
Fixes #298

### CODE REVIEW

File: datautils/jsonobject.go [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+// false unless the key exists
+func (d *JSONObject) HasKey(key string) bool {
+	return (*d)[key] != nil
+}
+
+func (d *JSONObject) ensureKeyExists(key string) {
+	if d.HasKey(key) {
+		return
+	}
+	panic(fmt.Errorf("No key found for %s", key))
```
Comment (aaron-collier) [inline datautils/jsonobject.go]: Do we want this to panic or return a more graceful error message? When I curl a POST that fails here I get stack trace back to the client, which feels ugly, but I don't think this should be a blocker here, just a question to possibly ticket later.

Comment (aaron-collier) [reply datautils/jsonobject.go]:
```
DN0a23334b:taco sul.amcollie$ curl -X POST -H "Content-Type: application/json" -d@examples/create-bs646cd8717.json http://localhost:8080/v1/resource -i
HTTP/1.1 100 Continue

HTTP/1.1 500 Internal Server Error
Date: Fri, 23 Mar 2018 16:37:20 GMT
Content-Type: text/plain; charset=utf-8
Connection: close
Transfer-Encoding: chunked

{"message":"No key found for id","stack":"goroutine 25 [running]:
runtime/debug.Stack(0xc420728000, 0x1f4, 0xc4201fe0f0)
	/usr/local/Cellar/go/1.9.3/libexec/src/runtime/debug/stack.go:24 +0xa7
github.com/sul-dlss-labs/taco/middleware.(*Recovery).ServeHTTP.func1(0x4c564e0, 0xc420728000, 0xc420202620)
	/Users/sul.amcollie/go/src/github.com/sul-dlss-labs/taco/middleware/recovery.go:38 +0x91
panic(0x46f6420, 0xc4201fe0f0)
	/usr/local/Cellar/go/1.9.3/libexec/src/runtime/panic.go:491 +0x283
github.com/sul-dlss-labs/taco/datautils.(*JSONObject).ensureKeyExists(0xc4205d6010, 0x47e3541, 0x2)
	/Users/sul.amcollie/go/src/github.com/sul-dlss-labs/taco/datautils/jsonobject.go:38 +0x113
github.com/sul-dlss-labs/taco/datautils.(*JSONObject).GetS(0xc4205d6010, 0x47e3541, 0x2, 0xc420257501, 0xc4205d6010)
	/Users/sul.amcollie/go/src/github.com/sul-dlss-labs/taco/datautils/jsonobject.go:10 +0x43
github.com/sul-dlss-labs/taco/datautils.(*Resource).ID(0xc4205d6010, 0xc4205d6010, 0xc4205d6008)
	/Users/sul.amcollie/go/src/github.com/sul-dlss-labs/taco/datautils/resource.go:55 +0x42
github.com/sul-dlss-labs/taco/db.DynamodbDatabase.Read(0xc4204cc028, 0x46e3e46, 0x9, 0xc420014820, 0x11, 0x9, 0x64, 0xc420596000)
	/Users/sul.amcollie/go/src/github.com/sul-dlss-labs/taco/db/database.go:52 +0x2cc
github.com/sul-dlss-labs/taco/db.(*DynamodbDatabase).Read(0xc4200e2b60, 0xc420014820, 0x11, 0x64, 0xc420312998, 0x0)
	<autogenerated>:1 +0x66
github.com/sul-dlss-labs/taco/validators.(*collectionStructuralValidator).ValidateResource(0xc4204e4990, 0xc420412028, 0x0, 0x0)
	/Users/sul.amcollie/go/src/github.com/sul-dlss-labs/taco/validators/collection_structural_validator.go:37 +0x13e
github.com/sul-dlss-labs/taco/validators.(*compositeResourceValidator).ValidateResource(0xc42063ab20, 0xc420412028, 0x0, 0x0)
	/Users/sul.amcollie/go/src/github.com/sul-dlss-labs/taco/validators/composite_resource_validator.go:16 +0x77
github.com/sul-dlss-labs/taco/validators.(*compositeResourceValidator).ValidateResource(0xc42063ab60, 0xc420412028, 0x0, 0xc420041910)
	/Users/sul.amcollie/go/src/github.com/sul-dlss-labs/taco/validators/composite_resource_validator.go:16 +0x77
github.com/sul-dlss-labs/taco/handlers.(*depositResource).Handle(0xc420402600, 0xc420536300, 0x4704240, 0xc4204ce600, 0xc420202800, 0x0)
	/Users/sul.amcollie/go/src/github.com/sul-dlss-labs/taco/handlers/deposit_resource.go:39 +0x99
github.com/sul-dlss-labs/taco/generated/restapi/operations.(*DepositResource).ServeHTTP(0xc42063bce0, 0x4c55960, 0xc4202027c0, 0xc420536300)
	/Users/sul.amcollie/go/src/github.com/sul-dlss-labs/taco/generated/restapi/operations/deposit_resource.go:56 +0x16b
github.com/sul-dlss-labs/taco/vendor/github.com/go-openapi/runtime/middleware.NewOperationExecutor.func1(0x4c55960, 0xc4202027c0, 0xc420536300)
	/Users/sul.amcollie/go/src/github.com/sul-dlss-labs/taco/vendor/github.com/go-openapi/runtime/middleware/operation.go:28 +0x6e
net/http.HandlerFunc.ServeHTTP(0xc42029b330, 0x4c55960, 0xc4202027c0, 0xc420536300)
	/usr/local/Cellar/go/1.9.3/libexec/src/net/http/server.go:1918 +0x44
github.com/sul-dlss-labs/taco/vendor/github.com/go-openapi/runtime/middleware.NewRouter.func1(0x4c55960, 0xc4202027c0, 0xc420536200)
	/Users/sul.amcollie/go/src/github.com/sul-dlss-labs/taco/vendor/github.com/go-openapi/runtime/middleware/router.go:74 +0x364
net/http.HandlerFunc.ServeHTTP(0xc420458e00, 0x4c55960, 0xc4202027c0, 0xc420536200)
	/usr/local/Cellar/go/1.9.3/libexec/src/net/http/server.go:1918 +0x44
github.com/sul-dlss-labs/taco/vendor/github.com/go-openapi/runtime/middleware.Redoc.func1(0x4c55960, 0xc4202027c0, 0xc420536200)
	/Users/sul.amcollie/go/src/github.com/sul-dlss-labs/taco/vendor/github.com/go-openapi/runtime/middleware/redoc.go:72 +0x94
net/http.HandlerFunc.ServeHTTP(0xc420364400, 0x4c55960, 0xc4202027c0, 0xc420536200)
	/usr/local/Cellar/go/1.9.3/libexec/src/net/http/server.go:1918 +0x44
github.com/sul-dlss-labs/taco/vendor/github.com/go-openapi/runtime/middleware.Spec.func1(0x4c55960, 0xc4202027c0, 0xc420536200)
	/Users/sul.amcollie/go/src/github.com/sul-dlss-labs/taco/vendor/github.com/go-openapi/runtime/middleware/spec.go:45 +0x73
net/http.HandlerFunc.ServeHTTP(0xc420364440, 0x4c55960, 0xc4202027c0, 0xc420536200)
	/usr/local/Cellar/go/1.9.3/libexec/src/net/http/server.go:1918 +0x44
github.com/sul-dlss-labs/taco/middleware.(*RequestLogger).ServeHTTP(0xc42036e1c0, 0x4c564e0, 0xc420728000, 0xc420536200)
	/Users/sul.amcollie/go/src/github.com/sul-dlss-labs/taco/middleware/logger.go:32 +0xb9
github.com/sul-dlss-labs/taco/middleware.(*Recovery).ServeHTTP(0xc420202620, 0x4c564e0, 0xc420728000, 0xc420536200)
	/Users/sul.amcollie/go/src/github.com/sul-dlss-labs/taco/middleware/recovery.go:49 +0x90
github.com/sul-dlss-labs/taco/vendor/github.com/honeybadger-io/honeybadger-go.(*Client).Handler.func1(0x4c564e0, 0xc420728000, 0xc420536200)
	/Users/sul.amcollie/go/src/github.com/sul-dlss-labs/taco/vendor/github.com/honeybadger-io/honeybadger-go/client.go:99 +0x94
net/http.HandlerFunc.ServeHTTP(0xc420202680, 0x4c564e0, 0xc420728000, 0xc420536200)
	/usr/local/Cellar/go/1.9.3/libexec/src/net/http/server.go:1918 +0x44
net/http.serverHandler.ServeHTTP(0xc420726270, 0x4c564e0, 0xc420728000, 0xc420536200)
	/usr/local/Cellar/go/1.9.3/libexec/src/net/http/server.go:2619 +0xb4
net/http.(*conn).serve(0xc42008dc20, 0x4c56f20, 0xc420298200)
	/usr/local/Cellar/go/1.9.3/libexec/src/net/http/server.go:1801 +0x71d
created by net/http.(*Server).Serve
	/usr/local/Cellar/go/1.9.3/libexec/src/net/http/server.go:2720 +0x288
","type":"error"}DN0a23334b:taco sul.amcollie$ 
```

Comment (jcoyne) [reply datautils/jsonobject.go]: That's not what I get:
```
$ curl -X POST -H "Content-Type: application/json" -d@examples/create-bs646cd8717.json http://localhost:8080/v1/resource -i
HTTP/1.1 100 Continue

HTTP/1.1 422 Unprocessable Entity
Content-Type: application/json
Date: Fri, 23 Mar 2018 17:46:50 GMT
Content-Length: 0
Connection: close
```

Comment (jcoyne) [reply datautils/jsonobject.go]: It looks like that error is fixed in https://github.com/sul-dlss-labs/taco/pull/352

**Comment by cmh2166** [comment]
I still am not a fan of the overly nested JSON hash within the Resource struct, but am willing to let it go. Needs a rebase, though, for final checks.

**Pull request merged**